pkg/controller/elasticsearch/driver/upscale_state.go

```diff
@@ -101,41 +122,102 @@ func (s *upscaleState) recordMasterNodeCreation() {
 	if s.isBootstrapped {
 		s.allowMasterCreation = false
 	}
+	s.recordNodesCreation(1)
 }
 
-func (s *upscaleState) canAddMasterNode() bool {
-	return s.allowMasterCreation
+func (s *upscaleState) canCreateMasterNode() bool {
+	return s.getMaxNodesToCreate(1) == 1 && s.allowMasterCreation
 }
 
-func (s *upscaleState) limitMasterNodesCreation(
-	actualStatefulSets sset.StatefulSetList,
-	ssetToApply appsv1.StatefulSet,
-) appsv1.StatefulSet {
-	if !label.IsMasterNodeSet(ssetToApply) {
-		return ssetToApply
+func (s *upscaleState) recordNodesCreation(count int32) {
+	s.recordedCreates += count
+}
+
+func (s *upscaleState) getMaxNodesToCreate(noMoreThan int32) int32 {
+	if s.createsAllowed == nil {
+		// unbounded, so allow all that was requested
+		return noMoreThan
 	}
 
-	targetReplicas := sset.GetReplicas(ssetToApply)
-	actual, alreadyExists := actualStatefulSets.GetByName(ssetToApply.Name)
-	actualReplicas := int32(0)
-	if alreadyExists {
-		actualReplicas = sset.GetReplicas(actual)
+	left := *s.createsAllowed - s.recordedCreates
+	if left < noMoreThan {
+		return left
+	}
+
+	return noMoreThan
+}
+
+func (s *upscaleState) limitNodesCreation(
+	actual appsv1.StatefulSet,
+	toApply appsv1.StatefulSet,
+) (appsv1.StatefulSet, error) {
+	if label.IsMasterNodeSet(toApply) {
+		return s.limitMasterNodesCreation(actual, toApply)
+	}
+
+	actualReplicas := sset.GetReplicas(actual)
+	targetReplicas := sset.GetReplicas(toApply)
+	if actualReplicas == targetReplicas {
+		return toApply, nil
+	}
+
+	if err := s.buildOnce(); err != nil {
```

Comment:
it feels a bit weird that we build the pointer receiver object in a function attached to that same pointer receiver. Maybe pass `s` as an argument to the function instead?

Reply:
Otherwise 👍 for restoring the lazy behaviour of the upscaleState! That should save some useless ES calls.

Reply:
Not sure if I got what you meant, is now better?